File: Algorithms and Data Structure/go/sorts/quick-sort/quicksort.go
package quick

// QuickSort uses the quick sort algorithm to sort an integer slice.
func sort(arr []int) {
	left := 0
	right := len(arr)

	if left < right {
		pivot := Partition(arr)
		sort(arr[:pivot])
		sort(arr[pivot+1:])
	}
}

// Partition ...
func Partition(arr []int) int {
	pivot := arr[len(arr)-1]
	i := -1

	for j := 0; j < len(arr)-1; j++ {
		if arr[j] < pivot {
			i++
			arr[i], arr[j] = arr[j], arr[i]
		}
	}
	i++
	arr[i], pivot = pivot, arr[i]
	return i
}

File: Custom/cpp/sorts/QuickSort.cpp
int Partition(vector<int> &arr, int from, int to) {

    int wall = from;

    for (int i = from; i < to; i++) {
        if (arr[i] <= arr[to]) {
            swap(arr[i],arr[wall]);
            wall++;
        }
    }

    swap(arr[to], arr[wall]);
    return wall;
}

void QuickSort(vector<int> &arr, int left, int right) {

    if (left < right){
        int wall = Partition(arr, left, right);
        QuickSort(arr, left, wall - 1);
        QuickSort(arr, wall + 1, right);
    }
}
